snapshotter/devmapper/pool_device.go

```diff
@@ -49,107 +44,132 @@ func NewPoolDevice(ctx context.Context, poolName, dataVolume, metaVolume string,
 
 	log.G(ctx).Infof("using dmsetup: %s", version)
 
-	poolPath := dmsetup.GetFullDevicePath(poolName)
+	dbpath := filepath.Join(config.RootPath, config.PoolName+".db")
+	poolMetaStore, err := NewPoolMetadata(dbpath)
+	if err != nil {
+		return nil, err
+	}
+
+	poolPath := dmsetup.GetFullDevicePath(config.PoolName)
 	if _, err := os.Stat(poolPath); err == nil {
-		log.G(ctx).Debugf("reloading existing pool '%s'", poolPath)
-		if err := dmsetup.ReloadPool(poolName, dataVolume, metaVolume, blockSizeSectors); err != nil {
-			return nil, errors.Wrapf(err, "failed to reload pool '%s'", poolName)
+		log.G(ctx).Debugf("reloading existing pool %q", poolPath)
+		if err := dmsetup.ReloadPool(config.PoolName, config.DataDevice, config.MetadataDevice, config.DataBlockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to reload pool %q", config.PoolName)
 		}
 	} else {
 		if !os.IsNotExist(err) {
-			return nil, errors.Wrapf(err, "failed to stat for '%s'", poolPath)
+			return nil, errors.Wrapf(err, "failed to stat for %q", poolPath)
 		}
 
 		log.G(ctx).Debug("creating new pool device")
-		if err := dmsetup.CreatePool(poolName, dataVolume, metaVolume, blockSizeSectors); err != nil {
-			return nil, errors.Wrapf(err, "failed to create thin-pool with name '%s'", poolName)
+		if err := dmsetup.CreatePool(config.PoolName, config.DataDevice, config.MetadataDevice, config.DataBlockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to create thin-pool with name %q", config.PoolName)
 		}
 	}
 
 	return &PoolDevice{
-		poolName: poolName,
-		devices:  make(map[string]int),
+		poolName: config.PoolName,
+		metadata: poolMetaStore,
 	}, nil
 }
 
-func (p *PoolDevice) CreateThinDevice(deviceName string, virtualSizeBytes uint64) (int, error) {
-	if _, ok := p.devices[deviceName]; ok {
-		return 0, errors.Errorf("device with name '%s' already created", deviceName)
+func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, virtualSizeBytes uint64) error {
+	deviceInfo := &DeviceInfo{
+		Name: deviceName,
+		Size: virtualSizeBytes,
 	}
 
-	// Create device, retry if device id is taken
-	deviceID, err := p.tryAcquireDeviceID(func(thinDeviceID int) error {
-		return dmsetup.CreateDevice(p.poolName, thinDeviceID)
+	// Create thin device and save metadata
+	err := p.metadata.AddDevice(ctx, deviceInfo, func(devID int) error {
+		return dmsetup.CreateDevice(p.poolName, devID)
 	})
 
 	if err != nil {
-		return 0, errors.Wrap(err, "failed to create thin device")
+		return err
 	}
 
-	p.devices[deviceName] = deviceID
-	if err := dmsetup.ActivateDevice(p.poolName, deviceName, deviceID, virtualSizeBytes, ""); err != nil {
-		return 0, errors.Wrap(err, "failed to activate thin device")
-	}
+	// Activate thin device
+	err = p.metadata.UpdateDevice(ctx, deviceName, func(info *DeviceInfo) error {
+		info.IsActivated = true
+		return dmsetup.ActivateDevice(p.poolName, info.Name, info.DeviceID, info.Size, "")
```

Comment:
How long does `dmsetup.ActivateDevice` take?  It looks like this is done within the scope of the write transaction which blocks all other access to the db.

Reply:
Created an issue to track this: https://github.com/firecracker-microvm/firecracker-containerd/issues/61